business/mesh.go

```diff
@@ -54,47 +57,49 @@ func NewMeshService(
 }
 
 func (in *MeshService) CanaryUpgradeStatus() (*models.CanaryUpgradeStatus, error) {
-	upgrade := in.conf.ExternalServices.Istio.IstioCanaryRevision.Upgrade
-	current := in.conf.ExternalServices.Istio.IstioCanaryRevision.Current
-	migratedNsList := []string{}
-	pendingNsList := []string{}
-
-	// If there is no canary configured, return empty lists
-	if upgrade == "" {
-		return &models.CanaryUpgradeStatus{MigratedNamespaces: migratedNsList, PendingNamespaces: pendingNsList}, nil
-	}
-
-	// Get migrated and pending namespaces
-	// TODO: Support multi-primary
-	migratedNss, err := in.homeClusterSAClient.GetNamespaces(fmt.Sprintf("istio.io/rev=%s", upgrade))
+	kubeCache, err := in.kialiCache.GetKubeCache(in.conf.KubernetesConfig.ClusterName)
 	if err != nil {
 		return nil, err
 	}
-	for _, ns := range migratedNss {
-		migratedNsList = append(migratedNsList, ns.Name)
-	}
 
-	pendingNss, err := in.homeClusterSAClient.GetNamespaces(fmt.Sprintf("%s=enabled", in.conf.IstioLabels.InjectionLabelName))
+	revisions, err := istio.GetHealthyIstiodRevisions(kubeCache, in.conf.IstioNamespace)
```

Comment:
Getting the revisions from the healthy pods has a side effect of returning zero revisions when the istiod has zero healthy pods. You can test this by creating a revision and then scaling the istiod replicas down to zero. This is probably an edge case but instead of getting the revisions from the pods, you can also use `in.discovery.Mesh().ControlPlanes`. This will return controlplanes across all clusters so you may need to then filter for a specific cluster if multi-primary isn't supported.

![Screenshot from 2024-08-26 15-07-35](https://github.com/user-attachments/assets/cd00736e-1b7f-4ed6-a2ec-e378e6134434)

Reply:
I had in mind to ignore unhealthy Istios from upgrading to, so retrieving healthy istiod pods were on purpose.
Maybe we could return back to this later based on business needs.